internal/sync/lifecycle.go

```diff
@@ -20,24 +20,111 @@
 
 package sync
 
+import (
+	"sync"
+
+	"go.uber.org/atomic"
+)
+
+// LifecycleState represents `states` that a lifecycle object can be in.
+type LifecycleState int
+
+const (
+	// Idle indicates the Lifecycle hasn't been operated on yet.
+	Idle LifecycleState = iota
+
+	// Starting indicates that the Lifecycle has begun it's "start" command
+	// but hasn't finished yet.
+	Starting
+
+	// Running indicates that the Lifecycle has finished starting and is
+	// available.
+	Running
+
+	// Stopping indicates that the Lifecycle 'stop' method has been called
+	// but hasn't finished yet.
+	Stopping
+
+	// Stopped indicates that the Lifecycle has been stopped.
+	Stopped
+
+	// Errored indicates that the Lifecycle experienced an error and we can't
+	// reasonably determine what state the lifecycle is in.
+	Errored
+)
+
 // LifecycleOnce is a helper for implementing transport.Lifecycles
 // with similar behavior.
 type LifecycleOnce struct {
-	start Once
-	stop  Once
+	lock sync.Mutex
+
+	state    atomic.Int32
+	startErr error
+	stopErr  error
 }
 
-// Start will run the `f` function once and return the error
+// Start will run the `f` function once and return the error.
+// If Start is called multiple times it will return the error
+// from the first time it was called.
 func (l *LifecycleOnce) Start(f func() error) error {
-	return l.start.Do(f)
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	// If we've already moved on from the idle state we've either
+	// called the start function already, or called the stop function
+	// in which case we should exit now and return the result of the
+	// last start command (or nil).
+	if LifecycleState(l.state.Load()) != Idle {
+		return l.startErr
+	}
+
+	// Set a nil function to an empty function
+	if f == nil {
+		f = func() error { return nil }
+	}
+
+	l.state.Store(int32(Starting))
+	l.startErr = f()
+	if l.startErr == nil {
+		l.state.Store(int32(Running))
+	} else {
+		l.state.Store(int32(Errored))
+	}
+
+	return l.startErr
 }
 
-// Stop will run the `f` function once and return the error
+// Stop will run the `f` function once and return the error.
+// If Stop is called multiple times it will return the error
+// from the first time it was called.
 func (l *LifecycleOnce) Stop(f func() error) error {
-	return l.stop.Do(f)
+	l.lock.Lock()
+	defer l.lock.Unlock()
+
+	// If the lifecycle is already stopped or errored, don't execute the stop func
+	if LifecycleState(l.state.Load()) == Stopped || LifecycleState(l.state.Load()) == Errored {
+		return l.stopErr
```

Comment:
if the state is errored because we called start, we need to return that error here instead

maybe we should use a single err attribute